crates/extensions/c8y_http_proxy/src/actor.rs

```diff
@@ -162,80 +165,146 @@ impl C8YHttpProxyActor {
         Ok(())
     }
 
-    async fn try_get_and_set_internal_id(&mut self) -> Result<(), C8YRestError> {
-        let internal_id = self.try_get_internal_id(None).await?;
-        self.end_point.set_c8y_internal_id(internal_id);
+    async fn try_get_and_set_internal_id(&mut self, device_id: String) -> Result<(), C8YRestError> {
+        if self
+            .end_point
+            .get_internal_id(Some(device_id.as_str()))
+            .is_none()
+        {
+            let internal_id = self.try_get_internal_id(device_id.clone()).await?;
+            self.end_point
+                .set_internal_id(self.end_point.device_id.clone(), internal_id);
+        }
         Ok(())
     }
 
-    async fn get_c8y_internal_child_id(
-        &mut self,
-        child_device_id: String,
-    ) -> Result<String, C8YRestError> {
-        if let Some(c8y_internal_id) = self.child_devices.get(&child_device_id) {
-            Ok(c8y_internal_id.clone())
-        } else {
-            let c8y_internal_id = self.try_get_internal_id(Some(&child_device_id)).await?;
-            self.child_devices
-                .insert(child_device_id, c8y_internal_id.clone());
-            Ok(c8y_internal_id)
+    async fn try_get_internal_id(&mut self, device_id: String) -> Result<String, C8YRestError> {
+        let url_get_id: String = self.end_point.get_url_for_get_id(device_id);
+        if self.end_point.token.is_none() {
+            self.get_fresh_token().await?;
         }
-    }
+        let build_request = |end_point: &C8yEndPoint| -> Result<HttpRequestBuilder, C8YRestError> {
+            Ok(HttpRequestBuilder::get(&url_get_id)
+                .bearer_auth(end_point.token.clone().unwrap_or_default()))
+        };
 
-    async fn try_get_internal_id(
-        &mut self,
-        device_id: Option<&str>,
-    ) -> Result<String, C8YRestError> {
-        let url_get_id = self.end_point.get_url_for_get_id(device_id);
+        let request_builder = build_request(&self.end_point)?;
+
+        let request = request_builder.build()?;
```

Comment:
I don't see the point to create and use the `build_request` closure in the same scope. Why not a simple method to build the request?

Reply:
This `closure` is not just used inside this function. When the first attempt fails because of `401/403` then the retries with the new `token`. Then it calls the `try_with_fresh_token`, this function takes the `closure. So, I created a closure here.